### ISSUE TITLE
bug: combination of external module loaded compiler and external depnedencies

### ISSUE THREAD

**Issue by JaroslavHron** [spack/spack#1787]
On a small cluster we use spack with compiler provided by external module:

compilers.yaml

``` yaml
- compiler:
    modules: [GCC/5.3.0-2.26]
    operating_system: centos6
    paths:
      cc: /apps/all/GCCcore/5.3.0/bin/gcc
      cxx: /apps/all/GCCcore/5.3.0/bin/g++
      f77: /apps/all/GCCcore/5.3.0/bin/gfortran
      fc: /apps/all/GCCcore/5.3.0/bin/gfortran
    spec: gcc@5.3.0
```

in combination with some external dependencies which I want to use from the system:

packages.yaml 

``` yaml
packages:
  all:
    compiler: [gcc@5.3.0]
    providers:
      mpi: [openmpi, mpich]
      blas: [openblas]
      lapack: [openblas]
  openssl:
    paths:
      openssl@system: /usr
    buildable: False
    version: [system]
```

this leads to a situation that the compilation environment setup by spack has
`/usr/bin` prepended to `PATH` (due to the requested external system openssl form packages.yaml)

However this leads to this cryptic assembler failure:

```
/tmp/cce6FYR4.s: Assembler messages:
/tmp/cce6FYR4.s:2159: Error: expecting string instruction after `rep'
make: *** [Parser/node.o] Error 1
```

The reason is that the external `gcc` module also loads corresponding libtools 
(which includes the correct `as`) 
but due to the prepended `/usr/bin` in `PATH` spack is picking `/usr/bin/as` instead.

Solutions might be:
1.  modify compiler wrappers so that the correct 'as' is used for example by  `gcc -B/path_to_as_bin_dir/`  and let user supply the `path_to_as_bin_dir` in compilers.yaml
2. or skip the systems dirs already in `PATH` when adding the external dependencies to the spack's `PATH`

Is there any better way to prevent this?

I have the 2. fix which helps and i can prepare it as PR.

**Comment by davydden** [comment]
i came across similar issues, see https://github.com/LLNL/spack/issues/1224 . 

(2) may brake some packages if the authors relied on some dependency be present in `PATH`, but i think it's the least evel in this case. One also needs to remove this path from CMake related variables.

**Comment by JaroslavHron** [comment]
I think that should not break any packages  - my reasoning is that the system path like `/usr/bin` is already in `PATH` anyway so the suggestion (2) (see  PR #1788) just prevents adding it again in-front of the other dirs loaded by modules.

I am not very familiar with CMake, but is there any specific `CMAKE_` variable as `PATH`?

**Comment by davydden** [comment]
> my reasoning is that the system path like /usr/bin is already in PATH anyway

you are right, i checked that it is indeed there already at the end of `PATH`.

> I am not very familiar with CMake, but is there any specific CMAKE_ variable as PATH?

Spack sets up `CMAKE_PREFIX_PATH` variable. The presence of system path there could also lead to surprises (and confuse CMake). The solution i could think of is to move system's path (if brought from external packages) at the very end.  @mathstuf  does this approach sounds reasonable to you?

**Comment by davydden** [comment]
p.s. @JaroslavHron you can check `build.env` in `.spack` of any installed package.

**Comment by JaroslavHron** [comment]
I see - `CMAKE_PREFIX_PATH` can also be a problem, but its slightly different case.

cmake seems to first search `CMAKE_PREFIX_PATH`, and then the internal variable `CMAKE_SYSTEM_PREFIX_PATH` (which holds the system dirs like ```CMAKE_SYSTEM_PREFIX_PATH= "/usr/local;/usr;/;/usr;/usr/local"```)

So the corresponding solution here would be to add to `CMAKE_PREFIX_PATH` only those dirs that are not already present in `CMAKE_SYSTEM_PREFIX_PATH`.

Right now i don't see a straight way (in python) to get the `CMAKE_SYSTEM_PREFIX_PATH` other than 
```cmake --system-information | grep CMAKE_SYSTEM_PREFIX_PATH```

**Comment by citibeth** [comment]
What happens if you just change `/usr` to `/fake/path` in your `packages.yaml` file?  Does that solve your problem?  See https://github.com/citibeth/spack/commit/ffeee437e04514e342f0d412371b2de52f2fbf42

**Comment by JaroslavHron** [comment]
Yes - using fake path in `packages.yaml` solves the problem. Thats good as quick fix.

**Comment by mathstuf** [comment]
Removing already-searched paths from `CMAKE_PREFIX_PATH` seems sensible. I don't see a better way than to use `--system-information`.

**Comment by becker33** [comment]
Closed by #23824
